### DISCUSSION TITLE
Interesting login issue.

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#126] [General]
Some time back, it happened, now it happened again.  I did an accidental double tap on return, mebbe 30 or so ms apart.  Both times AFTER entering the password.  The prompt "Password: " appears again, followed immediately by login.  I get occasional chatter if I am lax on kb cleaning, others probably do as well, if it happens to me...  Plus there is the general concept that a mainframe should not accept more than the authorized.  This could also (?) potentially lead to a buffer overflow issue(??). It accepts and acts upon a second actuation if quickly after the first, so it can be a potential issue with dirty keyboards, potentially elsewhere other than login...

I agree that dtcyber internet facing is not a likely hacker target, other than maybe some white hat out of curiosity (bring back the old days!), but buffer issues are bad no matter what.  I think it is more of a loop control thing at face value, but one would think the input would be ignored after the return until the shell takes over and all input between point a and b flushed into the bit bucket, still login is taking the cr after cr and responding with the pw prompt again, even though the password was submitted, and accepted (before or after that?)..  I'll look at the code later...loop control?

W10-recent, Haswell, Intel backend under vs22, as optimized as possible without error (IP, not IPO, etc....dtcyber and Intel multi-file optimization do not mix...yet)

What Cyber MAP dox are available?  OpenMP in July has three language standards on hold, possibly more....do concurrent, etc..(I digress, but...why the f not, if hlls and gpus get married in a July group wedding,..., even gaming gpus do how many sp teraflops, even gimped double on gaming cards is the better part of or over a teraflop now (the new commodities control list, hard limits, artificially introduced into the hardware)?)

Screenshot on request, but if it is happening on a fairly standard w10 corporate commodity laptop, I am probably not the first or only; this has happened twice for me.

As far as the subject matter of the post, I am wondering where the problem actually is, nos, dtcyber, or winblowz, but nos is doing the second prompt.

**Comment by kej715** [comment]
DtCyber itself does not interpret text entered by interactive users at
login (or, really, at any time). In particular, the emulated NPU/MDI does
not re-issue password prompts itself because it doesn't know the difference
between a password prompt or any other output generated by the operating
system running on the emulated mainframe. It seems that the behavior you've
observed can occur only if NVF, the NOS application that issues login
prompts and authenticates interactive users, is the thing that's
re-issuing the password prompt.

The NPU/MDI emulation in DtCyber simply sends upliine whatever
characters are received on the TCP connection, in the same way that a real
NPU/MDI would do it. It packages up the received text in formally-defined
PDU's (Protocol Data Units), and it sends them across the emulated channel
to the emulated PP. On interactive connections, reception of the <CR>
character triggers construction and transmission of an upline PDU. The
emulated NPU/MDI does the inverse for PDU's received across the emulated
channel. The upline and downline PDU streams are independent, and the
emulated NPU/MDI does not parse their contents on interactive terminal
connections.

Multiple <CR>'s received in quick sequence will simply cause the same
number of possibly empty PDU's to be constructed and sent upline over the
channel. There's no opportunity for buffer overflow. If NAM runs low on
memory for receiving and queueing PDU's, it will send flow control commands
to the NPU/MDI. If the NPU/MDI ignores the flow control requests, NAM will
stop the NPU/MDI and reload it.

What you're observing is almost certainly an unanticipated corner case
(i.e., bug) in NVF. Although it's unexpected and perhaps a bit annoying, it
should be harmless. It does not infer the possibility of buffer overflow or
security issue.

cheers,
Kevin

**Comment by KC5VDJ** [comment]
Yeah, I was thinking it would be such an edge case, it has only happened twice to me that I bothered noticing.

What do you think about the MAP idea?  They were an external option, but it's what cybers had, and I doubt anything using it other than realtime would use timing loops in a way that modern performance would harm anything, and with the new (and newly enhanced) GPU integration into the compilers by way of openmp.....

Thanks.
